### ISSUE TITLE
Feature Request: feature templates

### ISSUE THREAD

**Issue by Download** [developit/preact-cli#541]
**Do you want to request a *feature* or report a *bug*?**
Feature request

**What is the current behaviour?**
You create a project with the command:
```sh
preact create <template> <project-name>
```
The problem with this is that it leads to many, many templates if you want to easily create an app that supports certain features. For example, currently we have templates `material` and `widget`, but what if I want to create a Widget using Material UI?

**What is the expected behaviour?**
It would be great if we could create feature templates that would only add that feature to the current project i.s.o. defining the entire project from scratch. So let's say that we wanted to make a widget with Material Design. Widget might be a project template because it probably is different enough from a default app project to warrant that. But Material Design support probably just means adding some dependencies, pulling in a stylesheet and running some initialization code. So `material` could be a good candidate for a feature template. I would like to be able to do:

```sh
$ preact create widget my-widget with material
```
or
```sh
$ preact create widget my-widget
$ cd my-widget
$ preact add material
```

So basically I would like the syntax for `create` to be expanded to look like this:

```sh
preact create <project-template> <project-name> [with <feature-template> [, <feature-template>, ...]
```
and for an `add` command to be added that looks like this:

```sh
preact add <feature-template> [, <feature-template>, ...]
```

Where `preact create .. with` would just be a shortcut to `preact create ...  preact add ...`.

I think a feature template would look very similar to a project template, but would work on an existing project. Probably, a feature template would have to list the project templates it is compatible with.

Basically we would just apply the feature template on top of the project. 

**If this is a feature request, what is motivation or use case for changing the behaviour?**
I have been working on integrating [preact-i18nline](https://github.com/download/preact-i18nline) with Preact CLI. For now, this means a section in the README of `preact-i18nline` explaining how to adapt your (default) Preact project to use `preact-i18nline`. This README contains a configuration file among other things. So I thought it would be cool to create a project template for it... But then I was thinking that it sort of sucks to have to choose whether you want your app to support internationalization, or be a widget,, or use Material Design etc. We could of course make templates like `app-with-i18n`, `app-with-i18n-and-material`, `widget-with-i18n`, `widget-with-i18n-and-material` etc, but as you can already see you get an explosion of templates. A simple system to apply a feature template on top of a (compatible!) project template would really help here, so you could write:

```sh
preact create widget my-widget with i18n, material
```
I can imagine we would need to make some strict guidelines concerning directory layout etc. Maybe the project template could get some flag saying it adheres to these guidelines and should work with feature templates or something.

Thoughts?

**Comment by prateekbh** [comment]
This sounds like an ideal solution and I have given a thought to this in past, but the problem here might be that combining feature X and Y might start leading to conflicts. This might just make Feature devs life a lot tough.

e.g. Hypothetically redux might add something which AsyncRoute doesn't understand etc

Any solutions to contain this would be super excellent.

**Comment by Download** [comment]
If I understand the current system, Preact CLI basically copy-pastes the contents of the `template` folder into the new app folder, search-replacing the project name. Is that correct?

I agree features would be a lot more difficult to do 'perfect'. But we might get away by starting with something very simple and trying to discover where the issues lie. Then come up with a more robust solution for a v2.

I can imagine that just being able to 'augment' a skeleton project by basically taking the contents of a second template and merging it on top of the existing project could work surprisingly well. I guess we would need some special handling for package.json (read existing, then extend it with package.json from feature template, then write out the modified package.json) and (much harder probably) for *preact.config.js* but I guess we could come a long way with that alone. It would be enough for my project.

This seems like a good place for a small rant if you will forgive me but Webpack's *package.json* sucks. It's nice to be in JS when writing the config file, but for tools like Preact CLI it's hell. Merging a new package.json on top of an existing one is really, really simple. It's like 3 lines of code! But merging one webpack/preact.config.js on top of another??

Maybe it helps if I outline what I think I would need to make a `i18n` feature template? I started out thinking about this when I got the plan to create a `i18n` project template and realized it would mostly be a small patch on top of the `default` template. Maybe I should create that and we can see what exactly differs between the two and how hard it would be to patch just that difference?
